### PR TITLE
Bluetooth: GATT: Add function to get UUIDs for adv data

### LINKED ISSUE #33777: Bluetooth: GATT Service UUIDs in advertising data
**Is your enhancement proposal related to a problem? Please describe.**
When a server has certain capabilities expressed by the GATT services, it can add the UUIDs of those services to its advertising data, so that a central can determine the capabilities of the device before connection. Some GATT profiles even mandate that the service UUIDs are in the advertising data. 

Currently in Zephyr it is exclusively up to the application to manually add the service UUIDs to the advertising data. One minor issue with this approach is that Zephyr might enable some GATT services based on Kconfigs that aren't immediately obvious to the application, and thus the application may not know what service UUIDs it should advertise. 

**Describe the solution you'd like**
One solution to this is to have an API that the services can use to register their UUIDs (similar to how they register themselves with e.g. `bt_gatt_service_register`, or maybe even an extension of that). Then the UUIDs could either by appended to the advertising data automatically (perhaps with a `BT_LE_ADV_OPT_*`), or be returned in a separate function so that the application can trivially append the service UUIDs itself. 

**Describe alternatives you've considered**
It is technically possible to today, using the `bt_gatt_foreach_attr` API, to iterate on all the attributes (and thus the services) from the application to get all the UUIDs, so perhaps a wrapper/helper function on top of that might be a solution. 

**Additional context**
One challenge to consider is the length of all the UUIDs. If the device has several services and wants to advertise with e.g. legacy, then the number of UUIDs to add to the advertising data is quite low, and the advertising data type should perhaps be `BT_DATA_UUID16_SOME` rather than `BT_DATA_UUID16_ALL`. If only some of the UUIDs should be advertised, it is perhaps best if the application chooses which (i.e. not automaically append them). 
The solution should also take 32-bit UUID and 128-bit UUIDs into account.

### PR DESCRIPTION
**Pull request by Thalley**
Add a function that can help an application get get a
partial or complete list of UUIDs for advertisement purposes.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

fixes https://github.com/zephyrproject-rtos/zephyr/issues/33777

### CODE REVIEW

File: samples/bluetooth/peripheral/src/main.c [hunk 1]
```diff
@@ -265,6 +262,33 @@ static void bt_ready(void)
...
+		printk("Could not encode UUID128 data: %d\n", err);
+
+		err = bt_gatt_get_svc_uuid_data(BT_DATA_UUID128_SOME, &ad[2],
+						uuid128_buf,
+						sizeof(uuid128_buf));
+		if (err != 0) {
+			printk("Could not encode some UUID16 data: %d\n", err);
+			return;
+		}
+	}
```
Comment (Thalley) [inline samples/bluetooth/peripheral/src/main.c:290]: An alternative I considered was to automatically change the type to _SOME instead of _ALL if it did not fit, instead of returning an error and letting the application decided what to do.

File: subsys/bluetooth/host/gatt.c [hunk 2]
```diff
@@ -5696,3 +5697,153 @@ void bt_gatt_disconnected(struct bt_conn *conn)
...
+	struct bt_data *data;
+	uint8_t *data_buf;
+	const size_t data_buf_size;
+	bool truncated;
+};
+
+/* This is an array of UUID values that we return via the
+ * bt_gatt_get_svc_uuid_data for advertisements
+ */
+static uint16_t svc_uuid16_data_skip_list[] = {
```
Comment (alwa-nordic) [inline subsys/bluetooth/host/gatt.c:5711]: I suggest "exclude list" instead of "skip list". "skip list" can be confused with the type of data structure with that name.

File: subsys/bluetooth/host/gatt.c [hunk 3]
```diff
@@ -5696,3 +5697,153 @@ void bt_gatt_disconnected(struct bt_conn *conn)
...
 }
+
+struct svc_uuid_data_state {
+	struct bt_data *data;
+	uint8_t *data_buf;
+	const size_t data_buf_size;
+	bool truncated;
+};
+
+/* This is an array of UUID values that we return via the
```
Comment (alwa-nordic) [inline subsys/bluetooth/host/gatt.c:5708]: I think this should be "... that we don't return ...", right?

We should also add info about the selection criteria for the entries of this list. I suggest "List of Service UUIDs that are mandatory on a GATT server. Adverts for these Service UUIDs are redundant. bt_gatt_get_svc_uuid_data() filters out entries that are on this list."

File: include/bluetooth/gatt.h [hunk 4]
```diff
@@ -1773,6 +1773,35 @@ int bt_gatt_unsubscribe(struct bt_conn *conn,
  */
 void bt_gatt_cancel(struct bt_conn *conn, void *params);
 
+/** @brief Encode GATT UUID of current services into a bt_data struct.
+ *
```
Comment (alwa-nordic) [inline include/bluetooth/gatt.h:1777]: We should mention here that the mandatory GATT services are not included.

File: include/bluetooth/gatt.h [hunk 5]
```diff
@@ -1773,6 +1773,38 @@ int bt_gatt_unsubscribe(struct bt_conn *conn,
...
+ *                            results in.
+ * @param[out] data_buf       Pointer to the buffer where the encoded UUID data
+ *                            is stored.
+ * @param[out] data_buf_size  Size of the @p data_buf
+ *
+ * @retval 0 on success.
+ *
+ * @retval -EINVAL if any of the parameters are invalid (e.g. NULL or 0).
+ *
+ * @retval -ENOMEM if a _ALL type is used and @p data_buf does not fit all the
```
Comment (alwa-nordic) [inline include/bluetooth/gatt.h:1802]: The implementation leaves the out-parameters as if a _SOME type was used in the case or -ENOMEM. I can see users relying on this undocumented behavior. I'm not sure if it is a problem. (Is it?) We could deal with it now and specify this behavior in the API, or sabotage the output with memset (or at least set `data_len` to 0).

File: subsys/bluetooth/host/gatt.c [hunk 6]
```diff
@@ -5696,3 +5697,155 @@ void bt_gatt_disconnected(struct bt_conn *conn)
...
+		BT_DBG("data is NULL");
+		return -EINVAL;
+	}
+
+	CHECKIF(data_buf == NULL) {
+		BT_DBG("data_buf is NULL");
+		return -EINVAL;
+	}
+
+	data->type = type;
```
Comment (alwa-nordic) [inline subsys/bluetooth/host/gatt.c:5835]: This is a question, not a suggestion.. Should `data` be initialized / zeroed out here?

Comment (Thalley) [reply subsys/bluetooth/host/gatt.c:5832]: Not sure I completely follow. 

`bt_data` is a representation of a LTV structure. Resetting it here would not make much sense as we are overwriting all 3 values in the struct already.

File: include/bluetooth/gatt.h [hunk 7]
```diff
@@ -393,7 +393,18 @@ enum {
...
+ *   - @ref bt_uuid when UUID is @ref BT_UUID_GATT_PRIMARY or
+ *     @ref BT_UUID_GATT_SECONDARY.
+ *   - @ref bt_gatt_attr when UUID is @ref BT_UUID_GATT_INCLUDE.
+ *   - @ref bt_gatt_chrc when UUID is @ref BT_UUID_GATT_CHRC.
+ *   - _bt_gatt_ccc when UUID is @ref BT_UUID_GATT_CCC.
+ *   - @ref bt_gatt_cep when UUID is @ref BT_UUID_GATT_CEP.
+ *   - char * when UUID is @ref BT_UUID_GATT_CUD
+ *   - @ref bt_gatt_cpf when UUID is @ref BT_UUID_GATT_CPF
+ *
+ *  @param attr Attribute found. The
```
Comment (alwa-nordic) [inline include/bluetooth/gatt.h:407]:
```suggestion
 *  @param attr Attribute found.
```

File: samples/bluetooth/peripheral/src/main.c [hunk 8]
```diff
@@ -265,6 +262,33 @@ static void bt_ready(void)
...
 
+	err = bt_gatt_get_svc_uuid_data(BT_DATA_UUID16_ALL, &ad[1],
+					uuid16_buf, sizeof(uuid16_buf));
+	if (err != 0) {
+		printk("Could not encode all UUID16 data: %d\n", err);
+
+		err = bt_gatt_get_svc_uuid_data(BT_DATA_UUID16_SOME, &ad[1],
+						uuid16_buf, sizeof(uuid16_buf));
+		if (err != 0) {
+			printk("Could not encode some UUID16 data: %d\n", err);
```
Comment (asbjornsabo) [inline samples/bluetooth/peripheral/src/main.c:273]: "some" -> "any"?
Either able to encode all data, enable to encode some data, or not able to encode any data.
(The "some" is a bit tricky, if not able to encode all data, some data could not be encoded.  So "could not encode all" implies "could not encode some".)

Also elsewhere.

File: samples/bluetooth/peripheral/prj.conf [hunk 9]
```diff
@@ -16,10 +16,11 @@ CONFIG_BT_DEVICE_APPEARANCE=833
...
 CONFIG_BT_DEVICE_NAME_MAX=65
 
-CONFIG_BT_KEYS_OVERWRITE_OLDEST=y
-CONFIG_BT_SETTINGS=y
-CONFIG_FLASH=y
-CONFIG_FLASH_PAGE_LAYOUT=y
-CONFIG_FLASH_MAP=y
-CONFIG_NVS=y
-CONFIG_SETTINGS=y
+# CONFIG_BT_KEYS_OVERWRITE_OLDEST=y
```
Comment (asbjornsabo) [inline samples/bluetooth/peripheral/prj.conf:19]: The changes to this file seem unrelated to the rest of the commit.  Are they intended?

**Review by asbjornsabo — CHANGES_REQUESTED** [review]

File: include/bluetooth/gatt.h (hunk 5 above)
Comment (Thalley) [reply include/bluetooth/gatt.h:1818]: Done :)

File: samples/bluetooth/peripheral/src/main.c (hunk 8 above)
Comment (Thalley) [reply samples/bluetooth/peripheral/src/main.c:273]: I agree with your statement :) The reason behind using "some" here was that it was attempting to encode "some" (BT_DATA_UUID16_SOME) UUIDs. 

I can change it if you still feel like "any" is better

**Comment by Thalley** [comment]
Rebased to solve merge conflicts

**Review by MariuszSkamra — CHANGES_REQUESTED** [review]
I am not convinced this is right thing to do.

Most of the GATT Services don't need to to be advertised. There might be many circumstances under which service should not be advertised e.g. depending on privacy, whether the advertisement is connectable or not etc.
My feeling is that should be left to the user to build advertising data (like it is now).
Alternatively, `bt_gatt_service` could be extended with some callbacks that are called by some advertisement data builder, so that the service could decide whether it should be included in advertisement or not (depending on advertising type, privacy, etc).

**Comment by Thalley** [comment]
> 



> I am not convinced this is right thing to do.
> 
> Most of the GATT Services don't need to to be advertised. There might be many circumstances under which service should not be advertised e.g. depending on privacy, whether the advertisement is connectable or not etc. My feeling is that should be left to the user to build advertising data (like it is now). 

I don't disagree with this, which is why this is just a helper function to retrieve that data, and then it's up to the application to determine whether to include it all, or some. However, perhaps this should return an array of UUIDs instead of the data preformatted for adv data? 

> Alternatively, `bt_gatt_service` could be extended with some callbacks that are called by some advertisement data builder, so that the service could decide whether it should be included in advertisement or not (depending on advertising type, privacy, etc).

That seems to be the opposite of what you write above, isn't it? IIRC there is no requirement for any service to have it's UUID advertised. Some services/profile may, however, require that _service data_ (which is different from this) shall be advertised. For _service data_ I could see a benefit of such a callback.

**Comment by Thalley** [comment]
Rebased to solve merge conflict

**Comment by carlescufi** [comment]
@MariuszSkamra @asbjornsabo please re-review
@jori-nordic @alwa-nordic @MarekPieta @KAGA164 @jhedberg can you please take a look?

File: samples/bluetooth/peripheral/src/main.c (hunk 8 above)
Comment (Thalley) [reply samples/bluetooth/peripheral/src/main.c:295]: Changed to `Could not encode UUID16 data`, as we at this point won't have any encoded data.

File: include/zephyr/bluetooth/gatt.h [hunk 10]
```diff
@@ -1817,6 +1828,39 @@ int bt_gatt_unsubscribe(struct bt_conn *conn,
...
+ * @retval 0 on success.
+ *
+ * @retval -EINVAL if any of the parameters are invalid (e.g. NULL or 0).
+ *         In this case @p data_buf and @p data will be invalid.
+ *
+ * @retval -ENOMEM if a _ALL type is used and @p data_buf does not fit all the
+ *         data. In this case @p data_buf and @p data will be invalid.
+ */
+int bt_gatt_get_svc_uuid_data(uint8_t type, struct bt_data *data,
+			      uint8_t data_buf[], size_t data_buf_size);
```
Comment (jhedberg) [inline include/zephyr/bluetooth/gatt.h:1862]: Couldn't this API be simplified to have just two input parameters, if you require the caller to initialize the `bt_data` struct up-front with the type and buf values? Then the only additional thing you need to provide is the maximum length (i.e. size of the buffer).

You also either need to fix the documentation or the API name to use the same verb. In the documentation you say "encode" but in the function name you say "get".

File: samples/bluetooth/peripheral/src/main.c [hunk 11]
```diff
@@ -287,6 +284,33 @@ static void bt_ready(void)
...
+
+		err = bt_gatt_get_svc_uuid_data(BT_DATA_UUID128_SOME, &ad[2],
+						uuid128_buf,
+						sizeof(uuid128_buf));
+		if (err != 0) {
+			printk("Could not encode UUID128 data: %d\n", err);
+			return;
+		}
+	}
+
```
Comment (jhedberg) [inline samples/bluetooth/peripheral/src/main.c:313]: Is this really a good sample to choose to demonstrate the API? You end up adding many more lines and complexity compared to the original code. In fact, do you actually have a good use case for this API, where you end up with more readable code after the change?

**Review by jhedberg — CHANGES_REQUESTED** [review]
The general question I end up asking myself is that does this API really provide some added value over statically defining the advertising data (like the sample you're modifying used to do). I could imagine there maybe being some value for super complex and dynamic applications where the set of advertised services changes, and where the services are split into separate modules that you don't have direct visibility to (so querying the GATT layer that centrally manages these makes sense), but then we'd need to show that in the form of some sample application, rather than modifying one where this API doesn't really make the code any better (rather the opposite, IMO).

File: samples/bluetooth/peripheral/src/main.c (hunk 11 above)
Comment (Thalley) [reply samples/bluetooth/peripheral/src/main.c:313]: Yes, I have a better sample (unicast_audio_server) in mind that I can use instead :)

**Review by asbjornsabo — DISMISSED** [review]
My comments have been addressed

File: include/zephyr/bluetooth/gatt.h (hunk 10 above)
Comment (jhedberg) [reply include/zephyr/bluetooth/gatt.h:1862]: Just to add to this, if you insist on keeping the `type` and `data_buf` separate and only assigned to `bt_data` inside the call, then I would at least order the parameters so that `bt_data` becomes the first parameter,  Since that's really the only "out" parameter (I'm not sure it's right to call `data_buf` "out", since I don't think the caller is supposed to directly touch its content after the call, rather the "end product" of the call is the `bt_data` struct.

Comment (Thalley) [reply include/zephyr/bluetooth/gatt.h:1862]: I agree with your comments, and we can require that the user initializes the `bt_data` first with the `type` and `data` pointer. 

I'll try to make this change in a new commit (so it's easier to revert), and then I can just squash it if that's what we want.

**Review by jhedberg — APPROVED** [review]

File: include/zephyr/bluetooth/gatt.h (hunk 10 above)
Comment (Thalley) [reply include/zephyr/bluetooth/gatt.h]: @jhedberg Should I revert the commit to my original proposed API, or should we do something else? 

We could also, instead of using `struct bt_data`, simply put all UUID data in a `uint8_t` buffer, and then let the let the user put that buffer into their `struct bt_data`.

Comment (jhedberg) [reply include/zephyr/bluetooth/gatt.h]: Hmm, that's a tricky one. The reason it's const is to allow assignments to it from runtime evaluated literal values (like most samples are now doing).

> We could also, instead of using `struct bt_data`, simply put all UUID data in a `uint8_t` buffer, and then let the let the user put that buffer into their `struct bt_data`.

Yeah, that might be a cleaner solution, although the encoding format really only has a single user (advertising data). That said, the format is not special - just an array of consecutive values.

Comment (jhedberg) [reply include/zephyr/bluetooth/gatt.h]: Btw, if we go with this "encode into uint8_t array" solution  we also need to let the caller know if everything fit or not, and then let the caller then decide the advertising data type value. Or to put it in other words, when we remove the bt_data type from the API, it should not use BT_DATA_UUID* values as either input or output, to make it generic in clean way.

Comment (jhedberg) [reply include/zephyr/bluetooth/gatt.h]: ...it might be possible to convince me otherwise on that last point.. just thinking out loud. Let me know what you think

Comment (Thalley) [reply include/zephyr/bluetooth/gatt.h]: I'm also just thinking aloud here:

```c
 * @retval Number of octets put in @p data_buf.
 *
 * @retval -EINVAL if any of the parameters are invalid (e.g. NULL or 0).
 *         In this case @p data_buf and @p data will be invalid.
 *
 * @retval -ENOMEM if not all service UUIDs could be placed in @p data_buf. 
 *         @p data_buf will contain as many UUIDs as could be placed. In this 
 *         case @p data_buf can still be used with the BT_DATA_UUID16_SOME, 
 *         BT_DATA_UUID32_SOME or BT_DATA_UUID128_SOME depending on the @p type.
 */
int bt_gatt_get_svc_uuid_data(uint8_t type, uint8_t data_buf, size_t data_buf_size);
```

and where `type` would be one of BT_UUID_TYPE_16, BT_UUID_TYPE_32, BT_UUID_TYPE_128.

i.e. the function will pack as many service UUIDs in the `data_buf` as possible, and return the number of octets put in it. 

The returned buffer and size should be enough to then encode the `struct bt_data`. 

But like others, I do have doubt about the usefulness of this API, as A) it's pretty trivial to encode the service UUIDs yourself today, and it's quite difficult to provide a good API for this.

Comment (Thalley) [reply include/zephyr/bluetooth/gatt.h]: @jhedberg @MariuszSkamra What do you think? Is it worth pursuing this API, or just leave it?

Comment (jhedberg) [reply include/zephyr/bluetooth/gatt.h]: I'm ok leaving it. We can always revisit it later in the future if necessary (e.g. new use cases or users come along)

Comment (Thalley) [reply include/zephyr/bluetooth/gatt.h]: @jhedberg Should we leave it open, move it to draft or close it then? 

If we are not planning on merging it anytime soon, I'm in favor of just closing it.

Comment (jhedberg) [reply include/zephyr/bluetooth/gatt.h]: Closing is probably the best option, but maybe keep the branch in your fork so that this can be revisited later.

**Comment by Thalley** [comment]
Closing based on feedback. The idea may be revisited at some later point if we have more requests for it.
